neo/Consensus/ConsensusService.cs

```diff
@@ -153,12 +157,26 @@ private void Log(string message, LogLevel level = LogLevel.Info)
 
         private void OnChangeViewReceived(ConsensusPayload payload, ChangeView message)
         {
-            if (context.State.HasFlag(ConsensusState.CommitSent))
+            if (message.NewViewNumber < context.ViewNumber
+                // Change view requested to the same view needs to issue recovery if commit is sent, otherwise
```

Comment:
It is probably missing the case where `message.NewViewNumber > context.ViewNumber`.
In such case a node can have the `CommitFlag` set and `M` nodes had already achieved a higher view.
However, some of those `M` would had become byzantine after that. Thus, the node committed in a lower view can regenerate the others.

Reply:
Can you try to describe the scenario of how `M` nodes could count past the view number of the node that has `CommitSent` flag sent? If it has the commit flag sent then `M` nodes already had sent prepare responses; if those nodes restarted, they would have already been regenerated back to their previous state before their view gets back to the view number of the node that has the `CommitSent` flag set.

Now, let's suppose the node with `CommitSent` loses it's network connection and all other nodes restart. If they come to a consensus at a different or higher view number than the one stuck in `CommitSent` and a block is generated, the one with `CommitSent` flag will get unstuck when it accepts the block.

Reply:
`n = 2f +1`
No more than `f` nodes can be in different views with `CommitSent` flag set in the worst case.
With this code they would basically get restored to the view with the higher number that has `CommitSent` in such a case, and they would all commit in the higher view. I don't think there is an issue with the code not sending recovery to nodes at a higher view number.

Reply:
The case often happens and it is the one that current happens when the network gets stuck with two blocks.

* The node N_1 who is committed lost connection for some couple of seconds;
* The other nodes changed view;
* Some of them lost connection;
* Network is stuck in a higher view. Until nodes are back and perform a complete round of consensus or N_1 recovers them.

Reply:
I'll give it a bit more thought.

Reply:
There is an assumption that after CommitSent flag is on that node should not be modified or accept any modification.

Reply:
There is an assumption that after CommitSent flag is on that node should not be modified or accept any modification.
Which, strictly, should not be violated.

Reply:
> There is an assumption that after CommitSent flag is on that node should not be modified or accept any modification.
Which, strictly, should not be violated.

This code never modifies the state of a node with the `CommitSent` flag set. How is that related to this conversation?

Reply:
The one stuck with `CommitSent` flag doesn't get modified or accept modification, however if a new block is created that has signatures of `M` parties it of course will accept it and move on. There is nothing unsafe about that.

Reply:
So, to recap, it doesn't seem like it makes sense to recover higher nodes to lower view numbers, or else with two CN nodes with `CommitSent` at different views, the other nodes could get recovered back and forth between the two different view numbers. 

For that reason, I believe this current design is sufficient, as in that rare scenario the higher view number will be the one to which the other nodes get recovered, though it could require the other nodes to have to restart if they are poorly connected to the network and count their view past the one with the `CommitSent`. It is unlikely for them to do that if they are connected to nodes, since they should receive recovery messages, and if they do, they can just be restarted; it seems a reasonable trade-off compared to the potential of recovering back and forth to the different views repeatedly. 

What do you think? Can we resolve this issue?

Reply:
Jeff, not solved yet.
This is a requirement of the protocol, not an issue. aehuaheuaea

Considering the most simple case with 4 nodes: [n_1, ..., n_4]

1 - n_1 is committed and losses connection;
2 - n_2, n_3 and n_4 change to View_2;
3 - n_4 losses connection;
4 - n_1 is back.

This is not a rare situation. This is exactly what is happening today when we have two blocks for a same height.

The case you commented about moving back and forward will not happen, because:

* when a committed node (n_1) sends a recover package it contains all necessary information in its payload, enough for n_2 and n_3 automatically enter in the commit as well.
* if the network gets stuck and we count the occurred problems we gonna see a byzantine numbers of fault nodes.

Reply:
The case of moving back and forward will not happen with 4 CN nodes, you need to consider a case with more than 4 CN nodes to hit the problem I describe. Not all nodes may receive the recovery message sent at the same time. If there are 7 CN nodes and 2 different ones are in Commit at 2 different views, their recovery messages can be received in different orders by different nodes, and they could jump back and forth. I can give you more detail if the scenario if you like.

Reply:
After talking offline, I can agree that the case I’m describing shouldn’t happen because the recovery payload should include enough information for anyone to receive it to also commit immediately. I will change the PR so that once committed it will send recovery to any node that request change view even if it is coming from a higher view number.

Reply:
But, Jeff, after our conversation and your comments about the cases with 7 nodes, I also agree with you and was great to think about this again. 
It looks like we can have this case with N1, ..., N7:

* View 0
* N1 is committed and recover N3 (when asks change view);
* N2, N4, N5, N6 and N7 change view 1.
* Then, N2 commits and losses connection;
* N4, N5, N6 and N7 will try to change view (2) and will receive recovers from both N1 and N2;
* We can have a split between who recover who and stuck blocks generation with exactly, `f` fauty nodes (N1 and N2) and is prone to get stuck but not generate malicious blocks.

Maybe we need another additional mechanism, maybe we need to allow committed nodes to jump, only and strictly, to other committed phase with a higher view.

Reply:
I implemented allowing nodes to recover to lower views if `M` prepare requests are present now. However, that is actually wrong unless there are at least two Commits also, or more commits than itself if it is a node with commit sent.

Reply:
This is fully implemented now.

Reply:
I'm unresolving because I noticed on #426 @erikzhang commented that a node that had `CommitSent` should never change view; however, to fix the problem of jumping back and forth, @igormcoelho suggested that the node with `CommitSent` can change view if it receives a recovery with more signatures. That is what I implemented, but it violates @erikzhang design that he commented on in #426. 

I thought about it more and it leads to potential problems. Consider 7 Consensus nodes: A, B, C, D, E, F, G

A => Commit in 1
B => Commit in 3

Approximately simultaneously: 
* D => join B (Commit in 3) ; E=> join A (Commit in 1)
* C => send `ChangeView` and cause `RecoveryMessage` to get sent from D, and E

A => Commit in 3 (due to receive recovery from D, sees > commits than itself (since it still hasn't received the message from A joining yet), so it accepts recovery that has signatures from E+B
B => Commit in 1 (due to receive recovery from E, sees > commits than itself (since it hasn't received the message from D), so it accepts recovery that has signatures from D+A

This is now a problem because this is what the states of nodes now look like:
* A => Commit in 1 (sigs from A+E+B)
* B => Commit in 3 (sigs from B+D+A)

They have now got in a situation where they both have each other's signatures but they are in different views.

Finally to put the nail in the coffin.
Approximately simultaneously
* C => Commit in 1 (sigs from A+E+B+C) => generate block
* F => Commit in 3 (sigs from B+D+A+F) => generate block

And we have just created a block spork again.

I believe @erikzhang was right in saying that a node that has `CommitSent` cannot change view.

If we allow nodes to recover to lower view numbers, they will also join whichever commit it receives recovery from. In a 7 node system if, it could end up getting split 3 and 3, meaning the network would have to wait for the last node to come alive to break the tie. The system would essentially get stuck with < f nodes in a failed state, (1 < F=2). This is better than a block spork though.

Reply:
After commit phase change views concept should not exist anymore.

Reply:
In this sense, there are only committed nodes with different proposals.
The idea of a `Committed Node` change its view was new and is a little bit strange (I also hesitated, since it brokes that assumption I mentioned above about Committed nodes never being modified), but looks concise and a good metric if we use that idea Igor commented.

On the other hand, we should think with carefully. The only problem we saw was in that case with nodes getting split and network stuck with `f` faulty instead of `f+1`.

Reply:
The original issue for which this conversation was opened has been addressed. It will allow uncommitted nodes to now jump backward to join other committed nodes. However the criteria for jumping back currently is `M` preparations and 2 valid commit signatures.  Thinking about it more I think going back to accepting jumping back with `M` Preparations is better ignoring how many valid commit signatures there are, since ultimately that would allow it to more quickly join commits in a common view number and be able to converge and create the block.

Reply:
@jsolman, Jeff, as we discussed, the aforementioned example (counter-proof for the mechanism) needs an update because it is considering a block commit with 4 signatures.

Reply:
I'm going to update the earlier message now. and post back and the end here when updated.

Reply:
Perfect, Jeff. I will resolve this conversation because it main topic was accomplished in the last commits.

The recent discussions are related to an additional mechanism not completely necessary and analyzed in details yet.

Reply:
@vncoelh Please see my above example again; I updated it to fix the errors.

Reply:
The original issue for which this conversation was opened has been addressed. It will allow uncommitted nodes to now jump backward to join other committed nodes. The criteria for jumping back is `M` valid preparations seen in the recovery message. So I believe we can resolve this conversation now also.

Reply:
It was already resolved. 

I saw the example, Jeff. But looks like you have more than `f` different faulty.
Anyway, in the future we discuss that.

Reply:
But I also agree that this mechanism of allowing committed nodes to sign another committed block should not be used, it is not a strong assumption. 
It is better to let the network stuck in that specific situation we are trying to solve.
Currently, if we have two blocks we gonna be able to see who signed those two blocks are known which nodes are Malicious.